_posts/2018-08-14-apu2-validation-with-rte.md

```diff
@@ -0,0 +1,249 @@
+
+![Full setup](https://3mdeb.com/wp-content/uploads/2018/08/rte-apu-full-setup.jpg)
+
+## Theory of Operation
+
+Remote work assumes an active network connection, therefore the IP
+address of the RTE device should be checked beforehand. By default it's
+configured by DHCP, but I'm using RTE with static IP set to `192.168.3.105`. To
+establish a connection, type:
+
+```
+ssh root@192.168.3.105
+```
+
+and login as root user - in this example I'm suggesting to use Armbian OS, but
+another viable choice is Yocto RTE meta layer. Next open telnet connection to
+enable APU2 console preview. Check if ser2net redirection is configured:
+
+```
+root@orange-pi-zero:~# cat /etc/ser2net.conf
+13541:telnet:600:/dev/ttyS1:115200 8DATABITS NONE 1STOPBIT
+13542:telnet:600:/dev/ttyUSB0:115200 8DATABITS NONE 1STOPBIT
+```
+
+Open a new terminal or use `tmux` for split screen view and type:
+
+```
+telnet 192.168.3.105 13541
+```
+> we are using DB9 connector which is mapped to `/dev/ttyS1`
+
+Now, we have the ability to control Device Under Test from our personal computer
+and watch APU2 console on another tab.
+
+To test, if our connections are correct, we can go through a firmware flashing
+procedure. I will use mainline release v4.8.0.2 from PC Engines
+[github](https://pcengines.github.io/) site. Generally, it's an easy task.
+First, power on DUT by typing:
+
+```
+root@orange-pi-zero:~# echo 1 > /sys/class/gpio/gpio199/value
+```
+
+Save the APU2 console boot log for firmware version validation. In my example
+it is:
+
+```
+PC Engines apu2
+coreboot build 20180608
+BIOS version v4.8.0.1
+4080 MB ECC DRAM
+```
+
+Send target firmware image via ssh, e.g.
+
+```
+scp apu2_v4.8.0.2.rom root@192.168.3.105:/tmp/coreboot.rom
+```
+
+Run commands manually or create a shell script:
+
+```sh
+#!/bin/bash
+
+# power on APU2 platform with relay
+echo 1 > /sys/class/gpio/gpio199/value
+sleep 1
+
+# force the APU platform into ACPI S5 state
+echo 1 > /sys/class/gpio/gpio410/value
+sleep 5
+echo 0 > /sys/class/gpio/gpio410/value
+
+# flash APU ROM with flashrom
+flashrom -f -p linux_spi:dev=/dev/spidev1.0,spispeed=16000 -w /tmp/coreboot.rom
+
+# power on APU2 board
+echo 1 > /sys/class/gpio/gpio410/value
+sleep 1
+echo 0 > /sys/class/gpio/gpio410/value
+```
+
+Example flashrom output after flash verification:
+
+```
+flashrom v0.9.9-r1955 on Linux 4.17.2 (armv7l)
+flashrom is free software, get the source code at https://flashrom.org
+
+Calibrating delay loop... OK.
+Found Winbond flash chip "W25Q64.V" (8192 kB, SPI) on linux_spi.
+Reading old flash chip contents... done.
+Erasing and writing flash chip... Erase/write done.
+Verifying flash... VERIFIED.
+```
+
+New APU2 console boot log:
+
+```
+PC Engines apu2
+coreboot build 20180705
+BIOS version v4.8.0.2
+4080 MB ECC DRAM
+```
+
+You can see that platform BIOS changed from `v4.8.0.1` to the newer version
+`v4.8.0.2`, which confirms that flashing with RTE was successful :)
```

Comment:
@pietrushnic stosujemy emotki na blogu?

Reply:
@pietrushnic @miczyg1 Could you, please, review the article in a free moment?

Reply:
@mscibisz @pietrushnic I wonder about emoticons in the blog. I think it should be removed

Reply:
@miczyg1 I don't have any particular opinion - it would make thing more formal, but if it makes you feel better let's get rid of it.

Reply:
Well, all of our blog posts have abbreviations like `don't`, `I'm` etc. so they are written as "informal". Thus, it's not necessary to remove the smiley face.